### PR TITLE
Fix #8404: Use both scrutinee type and given type in pattern ascriptions

### LINKED ISSUE #3208: typed patterns get more precise types in scalac than dotty
The following code snippet fails to compile
```scala
class Test {
  trait E
  trait Marker

  def test(es: List[E]): List[E] = es.collect { case e: Marker => e }
}
```
```scala
-- Error: tests/allan/Test.scala:5:69 ------------------------------------------
5 |  def test(es: List[E]): List[E] = es.collect { case e: Marker => e }
  |                                                                     ^
  |Cannot construct a collection of type That
  |
  |where:    That is a type variable with constraint <: scala.collection.immutable.List[Test.this.E]
  | with elements of type Test.this.Marker based on a collection of type scala.collection.immutable.List[Test.this.E].
```

### LINKED ISSUE #8404: Don't use the declared type info when infer type in `match`
## minimized code

```Scala
trait ActorRefScope
class ActorRef
case class DeathWatchNotification(actor: ActorRef)
object Main {
  val ref: ActorRef = ???
  ref match {
    case r: ActorRefScope => DeathWatchNotification(r)
    case _ =>
  }
}
```

## Compilation output

```scala
[error] -- [E007] Type Mismatch Error: /dotty-example/src/main/scala/Main.scala:7:52 
[error] 7 |    case r: ActorRefScope => DeathWatchNotification(r)
[error]   |                                                    ^
[error]   |                                             Found:    (r : ActorRefScope)
[error]   |                                             Required: ActorRef
```

## expectation
Compile successfully as in scala 2

## info
dotty version: `0.22.0-RC1`

### PR DESCRIPTION
**Pull request by odersky**
In a pattern ascription `x: T`, type `x` to have the intersection of
the given type `T` and the scrutinee type. This matches scalac's behavior.
It now makes i5418.scala compile. The type intersection was enough to make
that one compile, it did not need unsound covariant GADT matching (which was
what I claimed before in the issue).

Fixes #3208

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -1573,8 +1573,20 @@ class Typer extends Namer
...
+            else if isWildcardStarArg(body1)
+                    || pt1 == defn.ImplicitScrutineeTypeRef
+                    || body1.tpe <:< pt  // There is some strange interaction with gadt matching.
+                                         // and implicit scopes.
+                                         // run/t2755.scala fails to compile if this subtype test is omitted
+                                         // and the else clause is changed to `body1.tpe & pt1`. What
+                                         // happens is that we get either an Array[Float] or an Array[T]
+                                         // where T is GADT constrained to := Float. But the case body
+                                         // compiles only if the bound variable is Array[Float]. If
+                                         // it is Array[T] we get an implicit not found. To avoid fragility
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: There's an open issue related to this that could be linked here: https://github.com/lampepfl/dotty/issues/5469

**Comment by allanrenucci** [comment]
I believe this fixes #3208 as well.

**Comment by odersky** [comment]
@allanrenucci 
> I believe this fixes #3208 as well.

yes, indeed!

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -1573,8 +1573,21 @@ class Typer extends Namer
...
+                                         // wrt to operand order for `&`, we include the explicit subtype test here.
+                                         // See also #5649.
+            then body1.tpe
+            else pt1 & body1.tpe
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: Do we need to check that `pt1.isValueType` to exclude potential ProtoTypes ?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: No, for pattern matching it is assured that `pt` is a value type.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: I added a link

**Pull request merged**